pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/EmptyResponseUtils.java

```diff
@@ -0,0 +1,206 @@
+      index++;
+    }
+    for (Pair<AggregationFunction, FilterContext> pair : filteredAggregationFunctions) {
+      AggregationFunction aggregationFunction = pair.getLeft();
+      columnNames[index] = AggregationFunctionUtils.getResultColumnName(aggregationFunction, pair.getRight());
+      columnDataTypes[index] = aggregationFunction.getFinalResultColumnType();
+      index++;
+    }
+    return new ResultTable(new DataSchema(columnNames, columnDataTypes), List.of());
+  }
+
+  /**
+   * Tries to fill an {@link DataSchema} when no row has been returned.
+   *
+   * Response data schema can be inaccurate (all columns set to STRING) when all segments are pruned on broker or
+   * server.
+   *
+   * Priority is:
+   * - Types from multi-stage engine validation for the given query (if allowed).
+   * - Types from schema for the given table (only applicable to identifiers).
+   * - Types from single-stage engine response (no action).
+   *
+   * Multi-stage engine schema will be available only if query compiles.
+   */
+  public static void fillEmptyResponseSchema(boolean useMSE, BrokerResponse response, TableCache tableCache,
+      Schema schema, String database, String query) {
+    Preconditions.checkState(response.getNumRowsResultSet() == 0, "Cannot fill schema for non-empty response");
+
+    DataSchema dataSchema = response.getResultTable() != null ? response.getResultTable().getDataSchema() : null;
+
+    List<RelDataTypeField> dataTypeFields = null;
+    // Turn on (with pinot.broker.use.mse.to.fill.empty.response.schema=true or query option
+    // useMSEToFillEmptyResponseSchema=true) only for clusters where no queries with huge IN clauses are expected
+    // (see https://github.com/apache/pinot/issues/15064)
+    if (useMSE) {
+      QueryEnvironment queryEnvironment = new QueryEnvironment(database, tableCache, null);
+      RelRoot root;
+      try (QueryEnvironment.CompiledQuery compiledQuery = queryEnvironment.compile(query)) {
+        root = compiledQuery.getRelRoot();
+      } catch (Exception ignored) {
+        root = null;
+      }
+      if (root != null && root.validatedRowType != null) {
+        dataTypeFields = root.validatedRowType.getFieldList();
+      }
+    }
+
+    if (dataSchema == null && dataTypeFields == null) {
+      // No schema available, nothing we can do
+      return;
+    }
+
+    if (dataSchema == null || (dataTypeFields != null && dataSchema.size() != dataTypeFields.size())) {
+      // If data schema is not available or has different number of columns than the validated row type, we use the
+      // validated row type to populate the schema.
+      int numColumns = dataTypeFields.size();
+      String[] columnNames = new String[numColumns];
+      ColumnDataType[] columnDataTypes = new ColumnDataType[numColumns];
+      for (int i = 0; i < numColumns; i++) {
+        RelDataTypeField dataTypeField = dataTypeFields.get(i);
+        columnNames[i] = dataTypeField.getName();
```

Comment:
Won't this use MSE style naming like `EXPR$0` for SSE queries? Are we fine with that?

Reply:
In most cases this code path shouldn't be triggered. It is only triggered when the column count doesn't match in SSE and MSE, and when user want to use MSE to fill the data schema.
This method is only moved and no logic is changed